tracer/src/Datadog.Trace/Debugger/ExceptionAutoInstrumentation/ILAnalyzer.cs

```diff
@@ -0,0 +1,117 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+
+#nullable enable
+namespace Datadog.Trace.Debugger.ExceptionAutoInstrumentation
+{
+    internal static class ILAnalyzer
+    {
+        public static bool HasDirectCallTo(MethodBase method, Type targetType, string methodName)
+        {
+            try
+            {
+                var effectiveMethod = GetEffectiveMethod(method);
+                return AnalyzeMethodBody(effectiveMethod, targetType, methodName);
+            }
+            catch
+            {
+                // If we can't analyze the method for any reason, assume it might not contain the call
+                return false;
+            }
+        }
+
+        private static MethodBase GetEffectiveMethod(MethodBase method)
+        {
+            // For async methods, analyze the MoveNext method
+            if (method.GetCustomAttribute<AsyncStateMachineAttribute>() is AsyncStateMachineAttribute asyncAttribute)
+            {
+                var stateMachineType = asyncAttribute.StateMachineType;
+                return stateMachineType.GetMethod(
+                           "MoveNext",
+                           BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public) ?? method;
+            }
+
+            return method;
+        }
+
+        private static bool AnalyzeMethodBody(MethodBase methodToAnalyze, Type targetType, string methodName)
+        {
+            var methodBody = methodToAnalyze.GetMethodBody();
+            if (methodBody == null)
+            {
+                return false;
+            }
+
+            var il = methodBody.GetILAsByteArray();
+            if (il == null)
+            {
+                return false;
+            }
+
+            int position = 0;
+            while (position < il.Length)
+            {
+                // Read opcode
+                int opcode = il[position++];
+                if (opcode == 0xFE && position < il.Length)
+                {
+                    opcode = 0xFE00 | il[position++];
+                }
+
+                // Check for call instructions
+                if (opcode == OpCodes.Call.Value || opcode == OpCodes.Callvirt.Value)
+                {
+                    // Ensure we have enough bytes for the token
+                    if (position + 3 >= il.Length)
+                    {
+                        break;
+                    }
+
+                    int token = il[position] |
+                               (il[position + 1] << 8) |
+                               (il[position + 2] << 16) |
+                               (il[position + 3] << 24);
+
+                    position += 4;
+
+                    try
+                    {
+                        var calledMethod = methodToAnalyze?.Module?.ResolveMethod(token);
```

Comment:
We can do that check outside the while loop for early exit

Reply:
We could, I don't really care though about this minor optimization

Reply:
Done in 4312d59534a010a7ad041ea21c1d9fb9e56ff78b.